parley/src/tests/utils/env.rs

```diff
@@ -0,0 +1,227 @@
+        self.check_counter += 1;
+        name
+    }
+
+    fn check_images(&self, current_img: &Pixmap, snapshot_path: &Path) -> Result<(), String> {
+        if !snapshot_path.is_file() {
+            return Err(format!("Cannot find snapshot {}", snapshot_path.display()));
+        }
+        let snapshot_img = Pixmap::load_png(snapshot_path)
+            .map_err(|_| format!("Loading snapshot {} failed", snapshot_path.display()))?;
+        if snapshot_img.width() != current_img.width()
+            || snapshot_img.height() != current_img.height()
+        {
+            return Err(format!(
+                "Snapshot has different size: snapshot {}x{}; generated image: {}x{}",
+                snapshot_img.width(),
+                snapshot_img.height(),
+                current_img.width(),
+                current_img.height()
+            ));
+        }
+
+        let mut n_different_pixels = 0;
+        let mut color_cumulative_difference = 0.0;
+        for (pixel1, pixel2) in snapshot_img.pixels().iter().zip(current_img.pixels()) {
+            if pixel1 != pixel2 {
+                n_different_pixels += 1;
+            }
+            let diff_r = (pixel1.red() as f32 - pixel2.red() as f32).abs();
+            let diff_g = (pixel1.green() as f32 - pixel2.green() as f32).abs();
+            let diff_b = (pixel1.blue() as f32 - pixel2.blue() as f32).abs();
+            color_cumulative_difference += diff_r.max(diff_g).max(diff_b);
+        }
+        if color_cumulative_difference > self.tolerance {
+            return Err(format!(
+                "Testing image differs in {n_different_pixels} pixels (color difference = {color_cumulative_difference})",
+            ));
+        }
+        Ok(())
+    }
+
+    pub(crate) fn check(&mut self, layout: &Layout<Color>) {
+        let current_img = render_layout(layout, self.background_color, self.foreground_color);
+        let image_name = self.image_name();
+
+        let snapshot_path = snapshot_dir().join(&image_name);
+        let comparison_path = current_imgs_dir().join(&image_name);
+
+        if let Err(e) = self.check_images(&current_img, &snapshot_path) {
+            if is_accept_mode() {
+                current_img.save_png(&snapshot_path).unwrap();
+            } else {
+                current_img.save_png(&comparison_path).unwrap();
+                self.errors.push((comparison_path, e));
+            }
+        }
+    }
+}
+
+impl Drop for TestEnv {
+    fn drop(&mut self) {
+        if !self.errors.is_empty() {
+            use std::fmt::Write;
+            let mut panic_msg = String::new();
+            for (path, msg) in &self.errors {
+                write!(
+                    &mut panic_msg,
+                    "{}\nImage written into: {}\n",
+                    msg,
+                    path.display()
+                )
+                .unwrap();
+            }
+            panic!("{}", &panic_msg);
```

Comment:
Does aborting if we're already panicking make sense here? I *think* it does, but I just want to gut check that.

Reply:
You are right. We should not panic when we are already panicking. I have fixed this.